### ISSUE TITLE
No apparent clean way to access the image object corresponding to a field

### ISSUE THREAD

**Issue by CAM-Gerlach** [lektor/lektor#623]
I posted this on the Lektor Gitter channel, and one of the devs who's worked on the image processing infrastructure, @xlotlu , instructed me to post this here as a bug.

I'm putting the finishing touches on a thoroughly overhauled version of Lektor-Icon, a modern, highly responsive single-page theme (of that sort that it seems all the flashy personal and business "showcase"-type sites are after these days..) that also includes a blog and other goodies. I'm wanting to implement responsive image sizing, i.e. to be able to feed lektor one high resolution source image and spit out differently-sized versions with ``thumbnail()`` for use with ``srcset``, or even just  resizing one image appropriately.

However, I'm pretty new to all this, so and I can't find a particularly clean and robust way to just turn the image name/path from a field into an image object that works with thumbnail.  It seems there should be some way either in the model setup or a function to get the image at a filename/path, but I can't seem to find it. The best I was able to do is call ``{% set hero_img_file = record.attachments.get(this.image) %}`` inside the HTML template for the flowblock, and then call ``.thumbnail()``. Here's a minimal example of such a flowblock model:

```
[block]
name = Hero

[fields.image]
label = Hero Background Image
type = select
source = record.attachments.images
```

and in my ``contents.lr`` I have

```
hero:

#### hero ####
image: camg-classic-storm-rainbow.jpg
---
```

But that seems quite clunky and overly complex; I figure there must be a better way I'm missing, but it seems @xlotlu didn't know of one either. Incidentally when trying to get the parent (to work around this and call ``this.parent.attachments``, before I stumbled across ``record.attachments``), I also found ``this.parent`` doesn't work for a flowblock; can that possibly be intended? Is there any other way to get a flowblock's parent object if its needed?

**Comment by xlotlu** [comment]
I didn't look closely at this, so initially I thought @CAM-Gerlach was using an "image"-type field, and he only got a string instead. That is of course not the case, lektor doesn't have such a field type.

While this seems an interesting idea, I'm not sure such an addition belongs in core, as it clashes with the concept of attachments somewhat. It does solve this particular problem, but the approach with lektor would normally be the other way around, to give attachments a model with a "is_hero" boolean (while ending up with a similarly clunky template, and not being able to enforce a single item, so not really solving the problem).

I think the real issue though is that `select` and `checkboxes` are field types, instead of field options. Take Django's `choices` for example -- it's something that applies to any field, and it's the callable's job to return something appropriate for that field type. This feels like the correct approach, and `choices` / `source` should be applicable to any type of field, and "multi" vs. "mono" would be just an extra-flag.

**Comment by nixjdm** [comment]
I do think there's a little confusion, and maybe on my part. Flow is intended to include an arbitrary number of the same kind of object in a page. So you're building a single-page site, it sounds like by having a single page with a number of "pages" created to scroll through by adding a bunch of content to the same real page - contents.lr. If you're using [block templating](https://www.getlektor.com/docs/content/flow/#block-templating) (which it sounds like), are you separating the images with anything? By adding template code to the block template files? That's my guess as to why you want to access the page you're on within the block templates.

You can access an image by name, just about as you would in the parent template. In the parent template you could have `this.attachments.get('hero1.png')`. In the block template - and this isn't documented well - you would instead have `record.attachments.get('hero1.png')` which would give you the same thing. `this` in the parent == `record` in the block template. That also means you don't need to try `this.parent` in a block template; it's just `record`.

All of that said, I'm suspicious you may want to refactor some things. Have you seen https://www.getlektor.com/docs/guides/single-page/ ? It achieves a single page site by stitching together other lektor pages, and uses no flowblocks. The advantage there, is that you can more easily break apart different sections of your "single page" into different real pages, which IMO is much easier to organize and keep track of. It'll end up giving you more manageable admin views and contents.lr files compared to many flow entries in one page. If each page now only has one image attachment since it's broken up, `{{ this.attachments.images.first() }}` would be sufficient, or a bit more complex mechanism if a child page is more complex - like grabbing the name from a field, or https://www.getlektor.com/plugins/lektor-thumbnail-generator/, or flow-blocks (on that smaller page).

Hope that helps!

**Comment by xlotlu** [comment]
@nixjdm he needs to select what particular image will be the hero out of all attachments, which seems like a reasonable expectation. (not a hardcoded name, or the first attachment)

**Comment by CAM-Gerlach** [comment]
Thanks so much @nixjdm for your detailed response and your work on Lektor! I'll reply to your questions point by point:

> Flow is intended to include an arbitrary number of the same kind of object in a page. 

I thought it was intended to accommodate more than one type of block; the overall high-level concept is essentially the similar to Lektor's very own homepage, which I see uses flowblocks in just that manner as the [docs](https://www.getlektor.com/docs/content/flow/) describe:

> As an example, the Front Page of this website uses many different blocks to implement the slideshow, feature list and more.

> So you're building a single-page site, it sounds like by having a single page with a number of "pages" created to scroll through by adding a bunch of content to the same real page - contents.lr. 

Technically, for clarity, I'm putting the finishing touches on an full overhaul of an existing Lektor theme, ported from a Hugo theme, ported from a plain HTML5 theme, that includes a "single-page"-style design for the mainpage but also includes other pages as well (and theoretically multiple "single-pages" can be added if the user desires). We currently use the original Lektor port (by another project member, more talented than I) as the basis for the [Spyder site](https://www.spyder-ide.org/), and the overhauled version still looks essentially similar in high-level design/layout (at least for the purposes of this question).

> If you're using block templating (which it sounds like)

Yep. The original Lektor port uses hardcoded flowblock fields for all the main section on the page, making it very inflexible—sections always appear in the same order, only one of each section can be present, they must be manually enabled/disabled in the project ini theme settings, etc.. I refactored this to use just one flowblock field for the page body, allowing the user to include an arbitrary number and order of flowblocks (and theoretically, their own custom section ones) all controlled by the user through the ``contents.lr`` on a per-page basis.

> are you separating the images with anything?

Unlike the other elements, there's only really intended to be one hero block (it is a hero after all :) ) at the top of the page, and currently it has its own dedicated flowblock field instead of being part of the main body flowblock field like all the others. It was originally a partial loaded with ``include``, which (to my understanding) doesn't allow its properties (background images, title/description text) to be set from the contents.lr file (and thus the admin panel) due to being in a separate inheritance hierarchy. and instead it was originally controlled globally from the theme settings in the project ini.

I could just inline it in the ``single-layout.html`` template, but I wanted to keep things clean and modular, although a flowblock field for a single block is a small layer of unneeded abstraction. Alternately, I could just include it as part of the body, but it is designed to go at the top (if present) and it makes it a little easier to access *from* the page (since I need to populate a ``<style>`` element in the page header block with the hero background image property in order to implement responsive images, which I can't do in an inline element ``style =``).

Is there a better way to handle that situation?

> That's my guess as to why you want to access the page you're on within the block templates.

Not exactly, at least for the purposes of this issue—as I mention above, its just so I can get at the actual image assets to automatically resize them with Lektor's ``thumbnail`` function in order to implement responsive image delivery (so users aren't downloading a 4K background image on a 720p phone screen where its only half-page-height).

> In the block template - and this isn't documented well - you would instead have record.attachments.get('hero1.png')

Yep, this is what I eventually figured out (as mentioned above), and I was wondering if there was a cleaner way that would just get the image object from the given path string (as you'd normally insert it on the page if you didn't have to do the thumbnail manipulation) without having to do the extra ``{% set image_obj = this/record.attachment.get(image_path) %}`` first every time. I later still ended up doing ``{% set hero_img_file = this.attachments.get(this.hero.blocks[0].image) %}`` on the main page instead so I could set the ``background_image`` property for the hero in a ``<style>`` element in the page head, since CSS responsive images can't be implemented in ``style= `` attributes and ``<style>`` is only allowed in ``<head>`` per the HTML spec (so I couldn't put it inside the hero block).

> Have you seen ``single-page``

Yes, actually; but I gave it a second look. I checked out the admin panel, and it does indeed look pretty long, although not (yet) unusably so. It certainly would be nice in principle to break things up for users into smaller chunks. However, in return for that usability boost it seems there would be a number of downsides at first glance, if I understand correctly:

* Users would not be able to create new single-page-layout pages from the admin panel and must know the specific steps to do so by hand
* Users would need to navigate to the correct subpage containing the main page items in the admin panel, and from there in turn to the appropriate sub-sub-page, rather than it all being inline in the order it actually is on the page
* There would be apparently no way to make the UI show the correct/valid blocks that can be added, since according to [the docs](https://www.getlektor.com/docs/models/children/) either the selection is restricted to one model, or all of them. Therefore, users would not necessarily know which could be added as valid blocks and which not.
* Ordering the blocks appears to be a bit of a problem; instead of the user being able to move blocks up and down at will and have the displayed order (in the admin panel or the contents.lr file) reflect that, it seems I'd need to implement either an ID or specialized property that the user would need to set to a certain string or integer on each block each and every time they wanted to update the order, which is not particularly quick, foolproof or intuitive

> he needs to select what particular image will be the hero out of all attachments

Yup, although to be clear I *am* able to do that; I'm just wondering if there is a cleaner way to implement it in the template without the extra ``{% set`` line and just getting the image object directly from a path, or even including an ``image`` field type that automatically treats its value as a path to an image (considering it works to do e.g. ``url{"{{ this.image_path | url }}")`` on the page itself, if I didn't have to manipulate the image object).

**Comment by CAM-Gerlach** [comment]
> just getting the image object directly from a path, or even including an image field type that automatically treats its value as a path to an image

This is essentially the primary request *this* issue was making from the beginning. My apologies if I got a bit off topic, but I was replying to @nixjdm 's questions and suggestions on this thread.

> @CAM-Gerlach the first half of this is captured in #625

Maybe I'm just new, but I can't see at all how ""select" / "checkboxes" should be field options, not field types" has anything to do with being able to "just get the image object directly from a path". There is no "image" or "path" etc. field type, so I just don't see how allowing ``select`` fields to be something other "string" would help at all given that none of the other types would to my knowledge make getting the image object from said filed any easier.

> . E.g. appending to a template block from another, to be able to push styles into <head>

If this is confirmed to be feasible and desired, I can go ahead with it. I just don't want to waste everyone's time creating an whole issue if the solution is either impossible, unlikely to be implemented, undesired or impractical.

**Comment by xlotlu** [comment]
> > @CAM-Gerlach the first half of this is captured in #625
> 
> Maybe I'm just new, but I can't see at all how ""select" / "checkboxes" should be field options, not field types" has anything to do with being able to "just get the image object directly from a path". There is no "image" or "path" etc. field type, so I just don't see how allowing `select` fields to be something other "string" would help at all given that none of the other types would to my knowledge make getting the image object from said filed any easier.

Indeed, "just get the image object directly from a path" only means "please give us an `image` field type". Your scenario however was "give me an image object from a selection" (the attachments' collection that is). Currently, `select` fields are not "strings", they are "select fields". And the problem lies in that their return object will always be a string, and not something else, since, well, they are not `something-else` fields.

> > . E.g. appending to a template block from another, to be able to push styles into
> 
> If this is confirmed to be feasible and desired, I can go ahead with it. I just don't want to waste everyone's time creating an whole issue if the solution is either impossible, unlikely to be implemented, undesired or impractical.

I personally find this desirable, but I have no idea how feasible it is (and if ever implemented, it's likely to happen in jinja, not lektor -- still, an issue submitted here will help prove its use-cases).

**Comment by xlotlu** [comment]
@CAM-Gerlach come to think of it, there could easily be be an `Image()` template function, and that might prove useful in other scenarios too (e.g. deriving dimensions or getting thumbnails of images from assets). @nixjdm what say you?

**Comment by CAM-Gerlach** [comment]
> Indeed, "just get the image object directly from a path" only means "please give us an image field type".

Well, not necessarily; that would be my other alternative that I didn't quote.What I meant with "just get the image object directly from a path" was what I think you're describing in your followup comment,

> there could easily be be an Image() template function

I.e. a function ``Image()`` that takes a string of the path and gets the appropriate image object, allowing the ``thumbnail()``, etc. methods and attributes to be called on it.

> And the problem lies in that their return object will always be a string

Right, I understood that part wrt adding the Image type to select fields and mentioned it in my previous post, but I mistakenly assumed you mean the "first part" of the two alternatives expressed in sentence you quoted (getting an Image from a string in the template), while I now see you actually mean the "first part" of implementing the second alternative.

**Comment by nixjdm** [comment]
> I thought it [flowblocks] was intended to accommodate more than one type of block

You're right. I was being reductionist, sorry. I was trying to emphasize the "arbitrary number" bit. Thank you for the background on what you're going through.

There's a lot I could address here, but I'll try to stay on topic.

> @CAM-Gerlach come to think of it, there could easily be be an Image() template function, and that might prove useful in other scenarios too (e.g. deriving dimensions or getting thumbnails of images from assets). @nixjdm what say you?

> I.e. a function Image() that takes a string of the path and gets the appropriate image object, allowing the thumbnail(), etc. methods and attributes to be called on it.

This is related to #484. One way or the other, assets, and in particular image assets, should be made easier to work with. That you can't call `.thumbnail` on an asset is problem. Soon, videos could have the same problem. Every method of Attachment instances and subclass (Images, Videos, future types) instances that makes sense for assets should still be something you can get, but for a lot of them you can't.
